src/script/miniscript.h

```diff
@@ -844,52 +868,164 @@ struct Node {
         assert(false);
     }
 
+    //! Get the net difference in the number of stack items after executing this fragment.
+    internal::NetDiff CalcNetDiff() const {
+        // B, K, and W fragments always push (or leave) exactly one element on the stack. V framgents never do.
+        const auto added_items{((typ & "V"_mst) == "V"_mst) ? 0 : 1};
+        // Get the maximum stack size necessary to (dis)satisfy this fragment as a negative number, if it exists.
+        // NOTE: it is fine to use the maximum stack size here, since we add the execution stack size value to the
+        // maximum initial stack size. Any increase in the execution stack size (by using a path that removes less items
+        // than we estimate here) necessarily involves an equal decrease in the initial stack size.
+        const auto neg_stack_size = [&](bool dsat) -> internal::MaxInt<int32_t> {
+            if (!dsat && ss.sat.size.valid) return -static_cast<int32_t>(ss.sat.size.value);
+            if (dsat && ss.dsat.size.valid) return -static_cast<int32_t>(ss.dsat.size.value);
+            return {};
+        };
+        return {added_items + neg_stack_size(/* dsat = */false), added_items + neg_stack_size(/* dsat = */true)};
+    }
+
     internal::StackSize CalcStackSize() const {
+        // Among non-malleable satisfactions and dissatisfactions, this computes for each fragment:
+        // - The maximum witness stack size of this (dis)satisfaction;
+        // - The maximum number of elements added to the stack during its execution. Note this computation is
+        // based on the net difference in stack size after executing fragments preceding it. This net difference
+        // is computed based on the maximum witness stack size of previous fragments' (dis)satisfactions.
         switch (fragment) {
-            case Fragment::JUST_0: return {{}, 0};
+            case Fragment::JUST_0: return {{}, {0, 1}};
             case Fragment::JUST_1:
             case Fragment::OLDER:
-            case Fragment::AFTER: return {0, {}};
-            case Fragment::PK_K: return {0, 0};
-            case Fragment::PK_H: return {1, 1};
+            case Fragment::AFTER: return {{0, 1}, {}};
+            case Fragment::PK_K: return {{0, 1}, {0, 1}};
```

Comment:
In 80606412f087eef74dedea008dc84a5efb715a49 "miniscript: check maximum stack size during execution"

I'm slightly confused by these return values. There will always be something on the stack for both satisfy and dissatisfy, and the fragment itself will push a key, so shouldn't this be `{1, 2}`?

Reply:
We don't account for what's already on the stack. We account for how many elements are *added* by this fragment (both during and after execution).

The check then consists in adding the maximum number of elements added to the stack and the maximum size of the initial witness stack: https://github.com/bitcoin/bitcoin/commit/80606412f087eef74dedea008dc84a5efb715a49#diff-a55760aaec4bce216663f5ebf65823516347356a8320d30459427149f7bbc2c5R1422-R1424.

Note this is (slightly) conservative, it's possible the spending path adding the largest number of elements to the stack is not the same as the one requiring the largest initial witness stack.